clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerRebalanceListenerInvoker.java

```diff
@@ -54,77 +64,102 @@ public class ConsumerRebalanceListenerInvoker {
         this.metricsManager = metricsManager;
     }
 
+    /**
+     * Invokes the onPartitionsAssigned callback method from the rebalance listener and logs the result.
+     *
+     * @param assignedPartitions the partitions assigned to the consumer
+     * @return an exception if an error occurred, or null if no error
+     */
     public Exception invokePartitionsAssigned(final SortedSet<TopicPartition> assignedPartitions) {
-        log.info("Adding newly assigned partitions: {}", assignedPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
-
-        Optional<ConsumerRebalanceListener> listener = subscriptions.rebalanceListener();
+        return invokeRebalanceCallback("Adding newly assigned partitions", assignedPartitions,
+                listener -> listener.onPartitionsAssigned(assignedPartitions),
+                metricsManager::recordPartitionsAssignedLatency);
+    }
 
-        if (listener.isPresent()) {
-            try {
-                final long startMs = time.milliseconds();
-                listener.get().onPartitionsAssigned(assignedPartitions);
-                metricsManager.recordPartitionsAssignedLatency(time.milliseconds() - startMs);
-            } catch (WakeupException | InterruptException e) {
-                throw e;
-            } catch (Exception e) {
-                log.error("User provided listener {} failed on invocation of onPartitionsAssigned for partitions {}",
-                        listener.get().getClass().getName(), assignedPartitions, e);
-                return e;
-            }
-        }
+    /**
+     * Invokes the onPartitionsRevoked callback method from the rebalance listener and logs the result.
+     *
+     * @param revokedPartitions the partitions revoked from the consumer
+     * @return an exception if an error occurred, or null if no error
+     */
+    public Exception invokePartitionsRevoked(final SortedSet<TopicPartition> revokedPartitions) {
+        return invokeRebalanceCallback("Revoke previously assigned partitions", revokedPartitions,
+                listener -> {
+                    Set<TopicPartition> revokePausedPartitions = subscriptions.pausedPartitions();
+                    revokePausedPartitions.retainAll(revokedPartitions);
+                    if (!revokePausedPartitions.isEmpty()) {
+                        log.info("The pause flag in partitions [{}] will be removed due to revocation.",
+                                revokePausedPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
+                    }
```

Comment:
What do you think about extracting this logic from `invokePartitionsRevoked()` and `invokePartitionsLost()` and making it a separate utility method? Something like:

```java
private void maybeLogPausedPartitions(SortedSet<TopicPartition> partitions, String reason) {
    if (!log.isInfoEnabled())
        return;

    Set<TopicPartition> pausedPartitions = subscriptions.pausedPartitions();
    pausedPartitions.retainAll(partitions);

    if (!pausedPartitions.isEmpty()) {
        log.info(
            "The pause flag in partitions [{}] will be removed due to {}.",
            reason
            pausedPartitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", "))
        );
    }
}
```